caddyhttp/digestauth/setup_test.go

```diff
@@ -0,0 +1,171 @@
+	}
+}
+
+func TestDigestAuthParse(t *testing.T) {
+	//htpasswdPasswd := "IedFOuGmTpT8"
+	htpasswdFile := `sha1:{SHA}dcAUljwz99qFjYR0YLTXx0RqLww=
+md5:$apr1$l42y8rex$pOA2VJ0x/0TwaFeAF9nX61`
+
+	var skipHtpassword bool
+	htfh, err := ioutil.TempFile(".", "digestauth-")
+	if err != nil {
+		t.Logf("Error creating temp file (%v), will skip htpassword test", err)
+		skipHtpassword = true
+	} else {
+		if _, err = htfh.Write([]byte(htpasswdFile)); err != nil {
+			t.Fatalf("write htpasswd file %q: %v", htfh.Name(), err)
+		}
+		htfh.Close()
+		defer os.Remove(htfh.Name())
+	}
+
+	tests := []struct {
+		input     string
+		shouldErr bool
+		password  string
+		expected  []Rule
+	}{
+		{`digestauth user pwd`, false, "pwd", []Rule{
+			{Users: NewSimpleUserStore(map[string]string{"user": "pwd"})},
+		}},
+		{`digestauth user pwd {
+		}`, false, "pwd", []Rule{
+			{Users: NewSimpleUserStore(map[string]string{"user": "pwd"})},
+		}},
+		{`digestauth /resource1 user pwd {
+		}`, false, "pwd", []Rule{
+			{Users: NewSimpleUserStore(map[string]string{"user": "pwd"}),
+				Resources: []string{"/resource1"}, Realm: "Restricted"},
+		}},
+		{`digestauth /resource1 user pwd {
+			realm Resources
+		}`, false, "pwd", []Rule{
+			{Users: NewSimpleUserStore(map[string]string{"user": "pwd"}),
+				Resources: []string{"/resource1"}, Realm: "Resources"},
+		}},
+		{`digestauth user pwd {
+			/resource1
+			/resource2
+		}`, false, "pwd", []Rule{
+			{Users: NewSimpleUserStore(map[string]string{"user": "pwd"}),
+				Resources: []string{"/resource1", "/resource2"}, Realm: "Restricted"},
+		}},
+		{`digestauth user pwd {
+			/resource1
+			/resource2
+			realm "Secure resources"
+		}`, false, "pwd", []Rule{
+			{Users: NewSimpleUserStore(map[string]string{"user": "pwd"}),
+				Resources: []string{"/resource1", "/resource2"}, Realm: "Secure resources"},
+		}},
+		{`digestauth user pwd {
+			/resource1
+			realm "Secure resources"
+			realm Extra
+			/resource2
+		}`, true, "pwd", []Rule{}},
+		{`digestauth user pwd {
+			/resource1
+			foo "Resources"
+			/resource2
+		}`, true, "pwd", []Rule{}},
+		{`digestauth /resource user pwd`, false, "pwd", []Rule{
+			{Users: NewSimpleUserStore(map[string]string{"user": "pwd"}), Realm: "Restricted"},
+		}},
+		{`digestauth /res1 user1 pwd1
+		  digestauth /res2 user2 pwd2`, false, "pwd", []Rule{
+			{Users: NewSimpleUserStore(map[string]string{"user1": "pwd1"}),
+				Resources: []string{"/res1"}, Realm: "Restricted"},
+			{Users: NewSimpleUserStore(map[string]string{"user2": "pwd2"}),
+				Resources: []string{"/res2"}, Realm: "Restricted"},
+		}},
+		{`digestauth user`, true, "", []Rule{}},
+		{`digestauth`, true, "", []Rule{}},
+		{`digestauth /resource user pwd asdf`, true, "", []Rule{}},
+
+		//{`digestauth sha1 htpasswd=` + htfh.Name(), false, htpasswdPasswd, []Rule{
+		//{Users: NewHtdigestUserStore(htfh.Name(), nil),
+		//Resources: []string{"/res2"}, Realm: "Restricted"},
+		//}},
```

Comment:
Why is this commented out?

Reply:
cause the tests are messed up. I tried to have "something" working as a start point. I left it for 2-3 weeks and had no idea what I wanted to do... :/